File: Task_3_Intelligent_Driver_Model_ACC/Acc_car/src/acc_control_cpp/src/acc_agent.cpp
#include "acc_agent.hpp"

#include <chrono>
#include <sstream>

using std::placeholders::_1;
using namespace std::chrono_literals;

/**
 * keep value within a range
 * @param value value to keen in range
 * @param min min value of range
 * @param max max value of range
 * @return value if value is in [min|max] or min or max
 */
template <typename T> T clamp(const T &value, const T &min, const T &max)
{
    return (value < min) ? min : (value > max) ? max : value;
}

/**
 * intelligent driver model
 * @see https://en.wikipedia.org/wiki/Intelligent_driver_model
 */
double intelligentDrivingModel(
    const double &desiredSpeed, const double &currentSpeed,
    const double &deltaSpeed, const double &currentGap,
    const double &maxAcceleration, const double &accelerationExponent,
    const double &headwayTime, const double &comfortableDeceleration,
    const double &safetyDist)
{

    auto accValue = maxAcceleration * (1.0 - pow(currentSpeed / desiredSpeed,
                                                 accelerationExponent));
    auto param_1 = headwayTime * currentSpeed;
    auto param_2 =
        (currentSpeed * deltaSpeed) /
        (2.0 * (pow(maxAcceleration * comfortableDeceleration, 0.5)));
    auto gap = fmax(safetyDist, param_1) + param_2;

    gap = (gap < 0.0) ? 0.0
                      : gap; // gap less than 0 - implies target faster than ego

    auto brakeValue = maxAcceleration * pow(gap / currentGap, 2);
    return accValue - brakeValue;
}

ACCAgent::ACCAgent()
    : Node("acc_agent_cpp"), m_radarDistance{-1}, m_gpsSpeed{-1}
{
    declare_parameter<std::string>("vehicle");
    m_vehicle_topic = get_parameter("vehicle").as_string();

    // ros parameter definition, can be used for tuining or hardcoded in
    // timer function
    declare_parameter<double>("max_speed", 0.0);
    m_maxSpeed = get_parameter("max_speed").as_double();
    declare_parameter<double>("max_acceleration", 0.0);
    m_maxAcceleration = get_parameter("max_acceleration").as_double();
    declare_parameter<double>("acceleration_exponent", 0.0);
    m_accelerationExponent = get_parameter("acceleration_exponent").as_double();
    declare_parameter<double>("headway_time", 0.0);
    m_headwayTime = get_parameter("headway_time").as_double();
    declare_parameter<double>("comfortable_deceleration", 0.0);
    m_comfortableDeceleration =
        get_parameter("comfortable_deceleration").as_double();
    declare_parameter<double>("safety_distance", 0.0);
    m_safetyDistance = get_parameter("safety_distance").as_double();

    set_parameter(
        rclcpp::Parameter("use_sim_time", rclcpp::ParameterValue(true)));

    m_currentValue = m_maxSpeed * 0.5;

    // ROS interface
    m_radarSubscription =
        create_subscription<webots_interfaces::msg::RadarTarget>(
            m_vehicle_topic + "/radar_targets",
            rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(
                            rmw_qos_profile_sensor_data),
                        rmw_qos_profile_sensor_data),
            std::bind(&ACCAgent::onRadar, this, _1));
    m_gpsSpeedSubscription = create_subscription<std_msgs::msg::Float32>(
        m_vehicle_topic + "/gps/speed",
        rclcpp::QoS(
            rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_sensor_data),
            rmw_qos_profile_sensor_data),
        std::bind(&ACCAgent::onGPSSpeed, this, _1));
    m_speedPublisher = create_publisher<std_msgs::msg::Float32>(
        m_vehicle_topic + "/cmd_speed", 1);

    m_timer = rclcpp::create_timer(this, get_clock(), 100ms,
                                   std::bind(&ACCAgent::onTimer, this));
    m_start = get_clock()->now();
}

void ACCAgent::onTimer()
{
    RCLCPP_INFO(get_logger(), "timer");
    if (get_clock().get()->now() -
            rclcpp::Time(0, 0, get_clock()->get_clock_type()) <
        rclcpp::Duration(5, 0))
    {
        // wait 5s for other vehicle to start
        return;
    }
    std_msgs::msg::Float32 s;
    // initalize with old value to make sure there is a value
    // if IDM cannot be called
    s.data = m_currentValue;
    float acc = 0.0f;

    // calculate speed value here
    // take care of old, therefore, invalid data

    RCLCPP_INFO(get_logger(),
                "value %f speed %f dist %f headway time %f idm %f", s.data,
                m_gpsSpeed, m_radarDistance, m_radarDistance / m_gpsSpeed, acc);

    // publish new pedal value
    m_currentValue = s.data;
    m_speedPublisher->publish(s);
}

void ACCAgent::onGPSSpeed(const std_msgs::msg::Float32::SharedPtr msg)
{
    // RCLCPP_INFO(get_logger(), "got gps speed %f", msg->data);
    if (msg->data < 500) // sometimes there is very big value as first package
    {
        m_gpsSpeed = msg->data;
        m_lastGPSSpeedUpdate = get_clock()->now();
    }
}

void ACCAgent::onRadar(const webots_interfaces::msg::RadarTarget::SharedPtr msg)
{
    // RCLCPP_INFO(get_logger(), "got radar dist %f", msg->distance);
    if (msg->distance > 2) // ghost in first curve
    {
        m_radarDistance = msg->distance;
        m_radarSpeed = msg->speed;
        m_lastRadarDistUpdate = get_clock()->now();
    }
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ACCAgent>());
    rclcpp::shutdown();
    return 0;
}

File: Task_3_Intelligent_Driver_Model_ACC/ACCAgent.py
import os
from re import S
import cv2
import numpy as np
import math
from pprint import pprint, pformat

import rclpy
from rclpy.duration import Duration
from rclpy.time import Time
from std_msgs.msg import Float32
from webots_interfaces.msg import RadarTarget
from rclpy.qos import qos_profile_sensor_data
from rclpy.node import Node
from rclpy.parameter import Parameter

def clamp(value, value_min, value_max):
    return min(max(value, value_min), value_max)

def intelligentDrivingModel(desiredSpeed, currentSpeed, deltaSpeed, currentGap,
                            maxAcceleration, accelerationExponent, headwayTime,
                            comfortableDeceleration, safetyDist):
    # https://en.wikipedia.org/wiki/Intelligent_driver_model
    accValue = maxAcceleration * \
        (1.0 - math.pow(currentSpeed / desiredSpeed, accelerationExponent))
    param_1 = headwayTime * currentSpeed
    param_2 = (currentSpeed * deltaSpeed) / (
        2.0 * (math.pow(maxAcceleration * comfortableDeceleration, 0.5)))
    gap = max(safetyDist, param_1) + param_2

    if gap < 0.0:  # gap less than 0 - implies target faster than ego
        gap = 0.0

    brakeValue = maxAcceleration * math.pow(gap / currentGap, 2)
    return accValue - brakeValue


class ACCAgent(Node):

    def __init__(self):
        super().__init__('acc_agent')

        self.declare_parameter("vehicle", "")
        self.vehicle_topic = self.get_parameter(
            'vehicle').get_parameter_value().string_value
        self.declare_parameter("max_speed", 75.0)
        self.max_speed = self.get_parameter(
            'max_speed').get_parameter_value().double_value
        self.declare_parameter("max_acceleration", 2.0)
        self.max_acceleration = self.get_parameter(
            'max_acceleration').get_parameter_value().double_value
        self.declare_parameter("acceleration_exponent", 4.0)
        self.acceleration_exponent = self.get_parameter(
            'acceleration_exponent').get_parameter_value().double_value
        self.declare_parameter("headway_time", 1.8)
        self.headway_time = self.get_parameter(
            'headway_time').get_parameter_value().double_value
        self.declare_parameter("comfortable_deceleration", 2.0)
        self.comfortable_deceleration = self.get_parameter(
            'comfortable_deceleration').get_parameter_value().double_value
        self.declare_parameter("safety_distance", 5.0)
        self.safety_distance = self.get_parameter(
            'safety_distance').get_parameter_value().double_value

        self.set_parameters(
            [Parameter('use_sim_time', Parameter.Type.BOOL, True)])

        # ROS interface
        self.__speed_publisher = self.create_publisher(
            Float32, self.vehicle_topic + '/cmd_speed', 1)
        self.create_subscription(RadarTarget,
                                 self.vehicle_topic + '/radar_targets',
                                 self.__on_radar, qos_profile_sensor_data)
        self.create_subscription(Float32, self.vehicle_topic + '/gps/speed',
                                 self.__on_gps_speed, qos_profile_sensor_data)

        self.radar_distance = -1
        self.last_radar_dist_update = None

        self.gps_speed = -1
        self.last_gps_speed_update = None

        self.current_value = self.max_speed * 0.5

        self.timer = self.create_timer(0.1, self.__timer_callback)

    def __on_radar(self, message):
        if message.distance > 2:  # ghost in first curve
            self.radar_distance = message.distance
            self.radar_speed = message.speed
            self.last_radar_dist_update = self.get_clock().now()

    def __on_gps_speed(self, message):
        if message.data < 500:  # sometimes there is a very big value as first package
            self.gps_speed = message.data
            self.last_gps_speed_update = self.get_clock().now()

    def __timer_callback(self):

        if self.get_clock().now() - Time(
                nanoseconds=0,
                clock_type=self.get_clock().clock_type) < Duration(seconds=5):
            return
        s = Float32()
        s.data = self.current_value
        acc = 0

        # calculate speed value here
        current_time = self.get_clock().now()
        radar_valid = self.last_radar_dist_update is not None and (current_time - self.last_radar_dist_update) < Duration(seconds=1)
        gps_valid = self.last_gps_speed_update is not None and (current_time - self.last_gps_speed_update) < Duration(seconds=1)
        # take care of old, therefore, invalid data
        if radar_valid and gps_valid and self.max_speed > 0:
            delta_speed = self.gps_speed - (self.gps_speed + self.radar_speed)
            acc = intelligentDrivingModel(
                desiredSpeed=self.max_speed,
                currentSpeed=self.gps_speed,
                deltaSpeed=delta_speed,
                currentGap=self.radar_distance,
                maxAcceleration=self.max_acceleration,
                accelerationExponent=self.acceleration_exponent,
                headwayTime=self.headway_time,
                comfortableDeceleration=self.comfortable_deceleration,
                safetyDist=self.safety_distance
            )
            s.data += acc * 0.1  # Update speed using acceleration and timer period (0.1s)
        else:
            # If data is invalid or max_speed is zero, apply gentle deceleration
            if self.gps_speed > 0:
                acc = -self.comfortable_deceleration * 0.5  # Gentle deceleration
            else:
                acc = 0
            s.data += acc * 0.1

        s.data = clamp(s.data, 0.0, self.max_speed)  # Clamp speed to valid range

        s.data = clamp(s.data, 0.0, self.max_speed)
        self.get_logger().info(
            "value %f speed %f dist %f headway time %f idm %f" %
            (s.data, self.gps_speed, self.radar_distance,
             self.radar_distance / self.gps_speed, acc))
        self.current_value = s.data
        self.__speed_publisher.publish(s)


def main(args=None):
    rclpy.init(args=args)
    agent = ACCAgent()
    rclpy.spin(agent)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
